src/sentry/api/endpoints/organization_stats_v2.py

```diff
@@ -0,0 +1,239 @@
+
+        result = outcomes.query(
+            start=start,
+            end=end,
+            rollup=rollup,
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [project.organization.id], "project_id": [project.id]},
+            orderby=["time"],
+        )
+        response = StatsResponse(start, end, rollup)
+        for row in result:
+            stat_to_update = response.get(row["category"])
+            stat_to_update.update(row)
+
+        return Response(response.build_fields())
+
+
+# TODO: good names for these classes
+class StatsResponse:
+    def __init__(self, start, end, rollup):
+        self.errors = TimeSeriesValues(start, end, rollup)
+        self.transactions = TimeSeriesValues(start, end, rollup)
+        self.attachments = TimeSeriesValues(start, end, rollup)
+
+    _GETTERS = {
+        DataCategory.ERROR: (lambda s: s.errors),
+        DataCategory.TRANSACTION: (lambda s: s.transactions),
+        DataCategory.ATTACHMENT: (lambda s: s.attachments),
+    }
+
+    ALL_FIELDS = frozenset(_GETTERS.keys())
+
+    def get(self, category):
+        return self._GETTERS[category](self)
+
+    def __iter__(self):
+        yield DataCategory.ERROR, self.errors
+        yield DataCategory.TRANSACTION, self.transactions
+        yield DataCategory.ATTACHMENT, self.attachments
+
+    def zerofill(self, start, end, rollup):
+        self.errors = zerofill(self.errors, start, end, rollup, "time")
+        self.transactions = zerofill(self.transactions, start, end, rollup, "time")
+        self.attachments = zerofill(self.attachments, start, end, rollup, "time")
+
+    def build_fields(self):
+        return {CATEGORY_NAME_MAP[category]: values.serialize() for category, values in self}
+
+
+# TODO: perhaps remove this class
+class TimeSeriesValues:
+    def __init__(self, start, end, rollup):
+        self.values = zerofill({}, start, end, rollup, "time")
+
+    def update(self, row):
+        self.values[row["time"]].update(row)
+
+    def serialize(self):
+        return [value.serialize() for value in self.values.values()]
+
+
+class StatMeasure:
+    def __init__(self, quantity=0, times_seen=0):
+        self.quantity = quantity
+        self.times_seen = times_seen
+
+
+class MeasureValue:
+    def __init__(self, time):
+        self.accepted = StatMeasure(0, 0)
+        self.filtered = StatMeasure(0, 0)
+        self.over_quota = StatMeasure(0, 0)
+        self.spike_protection = StatMeasure(0, 0)
+        self.other = StatMeasure(0, 0)
+        self.time = time
+
+    def serialize(self):
+        return {
+            "accepted": {
+                "quantity": self.accepted.quantity,
+                "times_seen": self.accepted.times_seen,
+            },
+            "filtered": {"quantity": self.filtered.quantity, "times_seen": self.filtered.quantity},
```

Comment:
Could you use the datacategory to decide whether to serialize quantity or times_seen as the output `value`? That would make frontend code simpler and reduce your payload size as we don't need both for each category.

Reply:
definitely.